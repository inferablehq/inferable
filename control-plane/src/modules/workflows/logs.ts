import { ulid } from "ulid";
import { events } from "../observability/events";

/**
 * Creates a log for a workflow execution.
 *
 * @param clusterId - ID of the cluster
 * @param workflowExecutionId - ID of the workflow execution
 * @param status - Status of the log (info, warn, error)
 * @param data - Additional data for the log
 * @returns Log info with ID, status, and creation timestamp
 */
export const createWorkflowLog = async ({
  clusterId,
  workflowExecutionId,
  status,
  data,
}: {
  clusterId: string;
  workflowExecutionId: string;
  status: "info" | "warn" | "error";
  data: Record<string, unknown>;
}) => {
  // Create an event for this notification
  const eventId = ulid();

  events.write({
    type: "workflowLogCreated",
    clusterId,
    jobId: workflowExecutionId,
    meta: {
      status,
      data,
    },
  });

  return {
    id: eventId,
    status,
    workflowExecutionId,
    createdAt: new Date(),
  };
};
