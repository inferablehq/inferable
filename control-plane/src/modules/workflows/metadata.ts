import { and, eq, desc } from "drizzle-orm";
import { db, workflowMetadata, workflows } from "../data";

export const getWorkflowsByMetadata = async ({
  clusterId,
  key,
  value,
  limit = 10,
  configId,
}: {
  clusterId: string;
  key: string;
  value: string;
  limit?: number;
  configId?: string;
}) => {
  return await db
    .select({
      id: workflows.id,
      name: workflows.name,
      userId: workflows.user_id,
      clusterId: workflows.cluster_id,
      status: workflows.status,
      createdAt: workflows.created_at,
      failureReason: workflows.failure_reason,
      debug: workflows.debug,
      test: workflows.test,
      configId: workflows.config_id,
      configVersion: workflows.config_version,
      feedbackScore: workflows.feedback_score,
    })
    .from(workflowMetadata)
    .where(
      and(
        eq(workflowMetadata.cluster_id, clusterId),
        eq(workflowMetadata.key, key),
        eq(workflowMetadata.value, value),
        ...(configId ? [eq(workflows.config_id, configId)] : []),
      ),
    )
    .rightJoin(workflows, eq(workflowMetadata.workflow_id, workflows.id))
    .orderBy(desc(workflows.created_at))
    .limit(limit);
};

export const getWorkflowMetadata = async ({
  clusterId,
  runId,
}: {
  clusterId: string;
  runId: string;
}) => {
  const metadata = await db
    .select({
      key: workflowMetadata.key,
      value: workflowMetadata.value,
    })
    .from(workflowMetadata)
    .where(
      and(
        eq(workflowMetadata.cluster_id, clusterId),
        eq(workflowMetadata.workflow_id, runId),
      ),
    );

  return metadata.reduce(
    (acc, { key, value }) => {
      acc[key] = value;
      return acc;
    },
    {} as Record<string, string>,
  );
};
