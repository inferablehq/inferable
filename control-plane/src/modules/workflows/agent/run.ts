import { DynamicStructuredTool } from "@langchain/core/tools";
import { z } from "zod";
import { NotFoundError } from "../../../utilities/errors";
import { getClusterContextText } from "../../cluster";
import { workflows } from "../../data";
import { embedSearchQuery } from "../../embeddings/embeddings";
import { flagsmith } from "../../flagsmith";
import { logger } from "../../observability/logger";
import {
  embeddableServiceFunction,
  getServiceDefinitions,
  serviceFunctionEmbeddingId,
} from "../../service-definitions";
import { getToolMetadata } from "../../tool-metadata";
import { notifyRunHandler } from "../notify";
import { getWorkflowMessages, insertRunMessage } from "../workflow-messages";
import { Run, getWaitingJobIds, updateWorkflow } from "../workflows";
import { createWorkflowAgent } from "./agent";
import { mostRelevantKMeansCluster } from "./nodes/tool-parser";
import { WorkflowAgentState } from "./state";
import {
  buildAbstractServiceFunctionTool,
  buildServiceFunctionTool,
} from "./tools/functions";
import {
  ACCESS_KNOWLEDGE_ARTIFACTS_TOOL_NAME,
  buildAccessKnowledgeArtifacts,
} from "./tools/knowledge-artifacts";
import { buildMockFunctionTool } from "./tools/mock-function";
import { events } from "../../observability/events";

const internalToolsMap: Record<
  string,
  (
    workflow: Run,
    toolCallId: string,
  ) => DynamicStructuredTool | Promise<DynamicStructuredTool>
> = {
  [ACCESS_KNOWLEDGE_ARTIFACTS_TOOL_NAME]: buildAccessKnowledgeArtifacts,
};

/**
 * Run a workflow from the most recent saved state
 **/
export const run = async (run: Run) => {
  logger.info("Running workflow");

  // Parallelize fetching additional context and service definitions
  const [additionalContext, serviceDefinitions] = await Promise.all([
    buildAdditionalContext(run),
    getServiceDefinitions({
      clusterId: run.clusterId,
    }),
    updateWorkflow({
      ...run,
      status: "running",
    }),
  ]);

  const allAvailableTools: string[] = [];
  const attachedFunctions = run.attachedFunctions ?? [];

  allAvailableTools.push(...attachedFunctions);

  serviceDefinitions.flatMap((service) =>
    (service.functions ?? []).forEach((f) => {
      // Do not attach additional tools if `attachedFunctions` is provided
      if (attachedFunctions.length > 0 || f.config?.private) {
        return;
      }

      allAvailableTools.push(
        serviceFunctionEmbeddingId({
          serviceName: service.name,
          functionName: f.name,
        }),
      );
    }),
  );

  const mockToolsMap: Record<string, DynamicStructuredTool> =
    await buildMockTools(run);

  const app = await createWorkflowAgent({
    workflow: run,
    allAvailableTools,
    additionalContext,
    getTool: async (toolCall) => {
      if (!toolCall.id) {
        throw new Error("Can not return tool without call ID");
      }

      const mockTool = mockToolsMap[toolCall.toolName];

      if (mockTool) {
        return mockTool;
      }

      const internalTool = internalToolsMap[toolCall.toolName];

      if (internalTool) {
        return internalTool(run, toolCall.id);
      }

      const serviceFunctionDetails = await embeddableServiceFunction.getEntity(
        run.clusterId,
        "service-function",
        toolCall.toolName,
      );

      if (serviceFunctionDetails) {
        return buildServiceFunctionTool({
          ...serviceFunctionDetails,
          schema: serviceFunctionDetails.schema,
          workflow: run,
          toolCallId: toolCall.id!,
        });
      }

      throw new NotFoundError(`Tool not found: ${toolCall.toolName}`);
    },
    findRelevantTools,
    postStepSave: async (state) => {
      logger.debug("Saving run state", {
        runId: run.id,
        clusterId: run.clusterId,
      });

      if (attachedFunctions.length == 0) {
        // optimistically embed the next search query
        // this is not critical to the workflow, so we can do it in the background
        embedSearchQuery(
          state.messages.map((m) => JSON.stringify(m.data)).join(" "),
        );
      }

      // Insert messages in a loop to ensure they are created with differing timestamps
      for (const message of state.messages.filter((m) => !m.persisted)) {
        await insertRunMessage(message);
        message.persisted = true;
      }
    },
  });

  const [messages, waitingJobIds] = await Promise.all([
    getWorkflowMessages({
      clusterId: run.clusterId,
      runId: run.id,
    }),
    getWaitingJobIds({
      clusterId: run.clusterId,
      runId: run.id,
    }),
  ]);

  try {
    const output = await app.invoke(
      {
        messages: messages.map((m) => ({
          ...m,
          persisted: true,
        })),
        waitingJobs: waitingJobIds,
        status: run.status,
      },
      {
        recursionLimit: 100,
      },
    );

    const parsedOutput = z
      .object({
        status: z.enum(workflows.status.enumValues),
        result: z.any().optional(),
        waitingJobs: z.array(z.string()),
      })
      .safeParse(output);

    if (!parsedOutput.success) {
      logger.error("Failed to parse workflow output", {
        parsedOutput,
      });
      throw new Error("Received unexpected workflow output state");
    }

    await updateWorkflow({
      ...run,
      status: parsedOutput.data.status,
    });

    const waitingJobs = parsedOutput.data.waitingJobs;

    await notifyRunHandler({
      run,
      status: parsedOutput.data.status,
      result: parsedOutput.data.result,
    });

    if (parsedOutput.data.status === "paused") {
      logger.info("Workflow paused", {
        waitingJobs,
      });

      return;
    }

    logger.info("Processing workflow complete");
  } catch (error) {
    logger.warn("Processing workflow failed", {
      error,
    });

    let failureReason = "An unknown error occurred during workflow processing.";
    if (error instanceof Error) {
      failureReason = error.message;
    }

    await updateWorkflow({
      ...run,
      status: "failed",
      failureReason,
    });

    throw error;
  }
};

async function findRelatedFunctionTools(workflow: Run, search: string) {
  const flags = await flagsmith?.getIdentityFlags(workflow.clusterId, {
    clusterId: workflow.clusterId,
  });

  const useKmeans = flags?.isFeatureEnabled("use_kmeans_clustering");

  const relatedToolsSearchResults = search
    ? await embeddableServiceFunction.findSimilarEntities(
        workflow.clusterId,
        "service-function",
        search,
        30,
      )
    : [];

  let relatedTools = relatedToolsSearchResults;

  if (useKmeans && relatedToolsSearchResults?.length > 1) {
    logger.info("Using kmeans clustering for related tool search");
    relatedTools = mostRelevantKMeansCluster(relatedToolsSearchResults);
  }

  const toolContexts = await Promise.all(
    relatedTools.map(async (toolDetails) => {
      const metadata = await getToolMetadata(
        workflow.clusterId,
        toolDetails.serviceName,
        toolDetails.functionName,
      );

      const contextArr = [];

      if (metadata?.additionalContext) {
        contextArr.push(`<context>${metadata.additionalContext}</context>`);
      }

      if (metadata?.resultKeys) {
        contextArr.push(
          `<result_keys>${metadata.resultKeys.slice(0, 10).map((k) => k.key)}</result_keys>`,
        );
      }

      return {
        serviceName: toolDetails.serviceName,
        functionName: toolDetails.functionName,
        toolContext: contextArr.join("\n\n"),
      };
    }),
  );

  const selectedTools = relatedTools.map((toolDetails) =>
    buildAbstractServiceFunctionTool({
      ...toolDetails,
      description: [
        toolDetails.description,
        toolContexts.find(
          (c) =>
            c?.serviceName === toolDetails.serviceName &&
            c?.functionName === toolDetails.functionName,
        )?.toolContext,
      ]
        .filter(Boolean)
        .join("\n\n"),
      schema: toolDetails.schema,
    }),
  );

  return selectedTools;
}

const buildAdditionalContext = async (workflow: Run) => {
  let context = "";

  context += await getClusterContextText(workflow.clusterId);
  context += `\nCurrent workflow URL: https://app.inferable.ai/clusters/${workflow.clusterId}/workflows/${workflow.id}`;
  workflow.systemPrompt && (context += `\n${workflow.systemPrompt}`);

  return context;
};

export const findRelevantTools = async (state: WorkflowAgentState) => {
  const workflow = state.workflow;

  const tools: DynamicStructuredTool[] = [];
  const attachedFunctions = workflow.attachedFunctions ?? [];

  // If functions are explicitly attached, skip relevant tools search
  if (attachedFunctions.length > 0) {
    for (const tool of attachedFunctions) {
      if (tool.toLowerCase().startsWith("inferable")) {
        const internalToolName = tool.split("_")[1];

        if (internalToolName === ACCESS_KNOWLEDGE_ARTIFACTS_TOOL_NAME) {
          tools.push(await buildAccessKnowledgeArtifacts(workflow));
          continue;
        }
      }

      const serviceFunctionDetails = await embeddableServiceFunction.getEntity(
        workflow.clusterId,
        "service-function",
        tool,
      );

      if (!serviceFunctionDetails) {
        throw new Error(
          `Tool ${tool} not found in cluster ${workflow.clusterId}`,
        );
      }

      tools.push(
        buildAbstractServiceFunctionTool({
          ...serviceFunctionDetails,
          schema: serviceFunctionDetails.schema,
        }),
      );
    }
  } else {
    const found = await findRelatedFunctionTools(
      workflow,
      state.messages.map((m) => JSON.stringify(m.data)).join(" "),
    );

    tools.push(...found);

    const accessKnowledgeArtifactsTool =
      await buildAccessKnowledgeArtifacts(workflow);

    // When tools are not specified, attach all internalTools
    tools.push(accessKnowledgeArtifactsTool);
  }

  return tools;
};

export const buildMockTools = async (workflow: Run) => {
  const mocks: Record<string, DynamicStructuredTool> = {};
  if (!workflow.testMocks || Object.keys(workflow.testMocks).length === 0) {
    return mocks;
  }

  if (!workflow.test) {
    logger.warn(
      "Workflow is not marked as test enabled but contains mocks. Mocks will be ignored.",
    );
    return mocks;
  }

  const serviceDefinitions = await getServiceDefinitions({
    clusterId: workflow.clusterId,
  });

  for (const [key, value] of Object.entries(workflow.testMocks)) {
    const [serviceName, functionName] = key.split("_");
    if (!serviceName || !functionName) {
      logger.warn("Invalid mock key", {
        key,
      });
      continue;
    }

    const mockResult = value.output;
    if (!mockResult) {
      logger.warn("Invalid mock output", {
        key,
        value,
      });
      continue;
    }

    const serviceDefinition = serviceDefinitions.find(
      (sd) => sd.name === serviceName,
    );

    if (!serviceDefinition) {
      logger.warn(
        "Service definition not found for mock. Mocks must refer to existing services.",
        {
          key,
          serviceName,
        },
      );
      continue;
    }

    const functionDefinition = serviceDefinition.functions?.find(
      (f) => f.name === functionName,
    );

    if (!functionDefinition) {
      logger.warn(
        "Function definition not found for mock. Mocks must refer to existing functions.",
        {
          key,
          serviceName,
          functionName,
        },
      );
      continue;
    }

    mocks[key] = buildMockFunctionTool({
      functionName,
      serviceName,
      description: functionDefinition.description,
      schema: functionDefinition.schema,
      mockResult,
    });
  }

  const mockKeys = Object.keys(mocks);
  if (mockKeys.length > 0) {
    logger.info("Built mock tools", {
      mockKeys,
    });
  }

  return mocks;
};
