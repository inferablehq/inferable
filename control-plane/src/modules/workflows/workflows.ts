import {
  InferSelectModel,
  and,
  countDistinct,
  desc,
  eq,
  inArray,
  ne,
  or,
  sql,
} from "drizzle-orm";
import { ulid } from "ulid";
import {
  BadRequestError,
  NotFoundError,
  RunBusyError,
} from "../../utilities/errors";
import { logger } from "../observability/logger";
import { Auth } from "../auth/auth";
import {
  clusters,
  db,
  jobs,
  RunMessageMetadata,
  workflowMessages,
  workflowMetadata,
  workflows,
} from "../data";
import {
  getWorkflowMessages,
  hasInvocations,
  lastAgentMessage,
  prepMessagesForRetry,
  upsertRunMessage,
} from "./workflow-messages";
import { env } from "../../utilities/env";
import { injectTraceContext } from "../observability/tracer";
import { getWorkflowMetadata } from "./metadata";
import { sqs } from "../sqs";
import { ChatIdentifiers } from "../models/routing";
import { customerTelemetry } from "../customer-telemetry";

export { start, stop } from "./queues";

export type Run = {
  id: string;
  clusterId: string;
  status?: "pending" | "running" | "paused" | "done" | "failed" | null;
  name?: string | null;
  configId?: string | null;
  systemPrompt?: string | null;
  failureReason?: string | null;
  debug?: boolean;
  test?: boolean;
  testMocks?: InferSelectModel<typeof workflows>["test_mocks"];
  feedbackComment?: string | null;
  feedbackScore?: number | null;
  resultSchema?: unknown | null;
  attachedFunctions?: string[] | null;
  onStatusChange?: string | null;
  interactive?: boolean;
  reasoningTraces?: boolean;
  enableSummarization?: boolean;
  modelIdentifier?: ChatIdentifiers | null;
  authContext?: unknown | null;
  context?: unknown | null;
};

export const createRun = async ({
  user,
  clusterId,
  name,
  test,
  testMocks,
  systemPrompt,
  onStatusChange,
  resultSchema,
  metadata,
  attachedFunctions,
  configId,
  configVersion,
  interactive,
  reasoningTraces,
  enableSummarization,
  modelIdentifier,
  customerAuthToken,
  authContext,
  context,
}: {
  user?: Auth;
  clusterId: string;
  name?: string;
  systemPrompt?: string;
  test?: boolean;
  testMocks?: Record<
    string,
    {
      output: Record<string, unknown>;
    }
  >;
  onStatusChange?: string;
  resultSchema?: unknown;
  metadata?: Record<string, string>;
  attachedFunctions?: string[];
  configId?: string;
  configVersion?: number;
  interactive?: boolean;
  reasoningTraces?: boolean;
  enableSummarization?: boolean;
  modelIdentifier?: ChatIdentifiers;
  customerAuthToken?: string;
  authContext?: unknown;
  context?: unknown;
}): Promise<Run> => {
  let run: Run | undefined = undefined;

  await db.transaction(async (tx) => {
    const [debugQuery] = await tx
      .select({
        debug: clusters.debug,
      })
      .from(clusters)
      .where(eq(clusters.id, clusterId));

    const result = await tx
      .insert(workflows)
      .values([
        {
          id: ulid(),
          cluster_id: clusterId,
          status: "pending",
          user_id: user?.entityId ?? "SYSTEM",
          ...(name ? { name } : {}),
          debug: debugQuery.debug,
          system_prompt: systemPrompt,
          test,
          test_mocks: testMocks,
          reasoning_traces: reasoningTraces,
          interactive: interactive,
          enable_summarization: enableSummarization,
          on_status_change: onStatusChange,
          result_schema: resultSchema,
          attached_functions: attachedFunctions,
          config_id: configId,
          config_version: configVersion,
          model_identifier: modelIdentifier,
          customer_auth_token: customerAuthToken,
          auth_context: authContext,
          context: context,
        },
      ])
      .returning({
        id: workflows.id,
        name: workflows.name,
        clusterId: workflows.cluster_id,
        systemPrompt: workflows.system_prompt,
        status: workflows.status,
        debug: workflows.debug,
        test: workflows.test,
        attachedFunctions: workflows.attached_functions,
        modelIdentifier: workflows.model_identifier,
        authContext: workflows.auth_context,
        context: workflows.context,
        interactive: workflows.interactive,
      });

    run = result[0];

    if (!!run && metadata) {
      await tx.insert(workflowMetadata).values(
        Object.entries(metadata).map(([key, value]) => ({
          cluster_id: clusterId,
          workflow_id: run!.id,
          key,
          value,
        })),
      );
    }
  });

  if (!run) {
    throw new Error("Failed to create run");
  }

  return run;
};

export const deleteRun = async ({
  clusterId,
  runId,
}: {
  clusterId: string;
  runId: string;
}) => {
  await db
    .delete(workflows)
    .where(and(eq(workflows.cluster_id, clusterId), eq(workflows.id, runId)));
};

export const updateWorkflow = async (workflow: Run): Promise<Run> => {
  if (workflow.status && workflow.status !== "failed") {
    workflow.failureReason = null;
  }

  const [updated] = await db
    .update(workflows)
    .set({
      name: !workflow.name ? undefined : workflow.name,
      status: workflow.status,
      failure_reason: workflow.failureReason,
      feedback_comment: workflow.feedbackComment,
      feedback_score: workflow.feedbackScore,
    })
    .where(
      and(
        eq(workflows.cluster_id, workflow.clusterId),
        eq(workflows.id, workflow.id),
      ),
    )
    .returning({
      id: workflows.id,
      name: workflows.name,
      clusterId: workflows.cluster_id,
      status: workflows.status,
      failureReason: workflows.failure_reason,
      debug: workflows.debug,
      attachedFunctions: workflows.attached_functions,
      authContext: workflows.auth_context,
      context: workflows.context,
    });

  // Send telemetry event if feedback was updated
  if (workflow.feedbackScore !== undefined && workflow.feedbackScore !== null) {
    customerTelemetry.track({
      type: "runFeedback",
      runId: workflow.id,
      clusterId: workflow.clusterId,
      score: workflow.feedbackScore,
      comment: workflow.feedbackComment || undefined,
    });
  }

  return updated;
};

export const getWorkflow = async ({
  clusterId,
  runId,
}: {
  clusterId: string;
  runId: string;
}) => {
  const [workflow] = await db
    .select({
      id: workflows.id,
      name: workflows.name,
      userId: workflows.user_id,
      configId: workflows.config_id,
      clusterId: workflows.cluster_id,
      systemPrompt: workflows.system_prompt,
      status: workflows.status,
      failureReason: workflows.failure_reason,
      debug: workflows.debug,
      test: workflows.test,
      testMocks: workflows.test_mocks,
      onStatusChange: workflows.on_status_change,
      resultSchema: workflows.result_schema,
      feedbackComment: workflows.feedback_comment,
      feedbackScore: workflows.feedback_score,
      attachedFunctions: workflows.attached_functions,
      reasoningTraces: workflows.reasoning_traces,
      interactive: workflows.interactive,
      enableSummarization: workflows.enable_summarization,
      modelIdentifier: workflows.model_identifier,
      authContext: workflows.auth_context,
      context: workflows.context,
    })
    .from(workflows)
    .where(and(eq(workflows.cluster_id, clusterId), eq(workflows.id, runId)));

  return workflow;
};

export const getClusterWorkflows = async ({
  clusterId,
  userId,
  test,
  limit = 50,
  configId,
}: {
  clusterId: string;
  test: boolean;
  userId?: string;
  limit?: number;
  configId?: string;
}) => {
  const result = await db
    .select({
      id: workflows.id,
      name: workflows.name,
      userId: workflows.user_id,
      clusterId: workflows.cluster_id,
      systemPrompt: workflows.system_prompt,
      status: workflows.status,
      createdAt: workflows.created_at,
      failureReason: workflows.failure_reason,
      debug: workflows.debug,
      test: workflows.test,
      configId: workflows.config_id,
      configVersion: workflows.config_version,
      feedbackScore: workflows.feedback_score,
      modelIdentifier: workflows.model_identifier,
      authContext: workflows.auth_context,
      context: workflows.context,
    })
    .from(workflows)
    .where(
      and(
        eq(workflows.cluster_id, clusterId),
        eq(workflows.test, test),
        ...(userId ? [eq(workflows.user_id, userId)] : []),
        ...(configId ? [eq(workflows.config_id, configId)] : []),
      ),
    )
    .orderBy(desc(workflows.created_at))
    .limit(limit);

  return result;
};

export const getWorkflowDetail = async ({
  clusterId,
  runId,
}: {
  clusterId: string;
  runId: string;
}) => {
  const [[workflow], agentMessage, metadata] = await Promise.all([
    db
      .select({
        id: workflows.id,
        name: workflows.name,
        userId: workflows.user_id,
        clusterId: workflows.cluster_id,
        status: workflows.status,
        systemPrompt: workflows.system_prompt,
        failureReason: workflows.failure_reason,
        debug: workflows.debug,
        test: workflows.test,
        feedbackComment: workflows.feedback_comment,
        feedbackScore: workflows.feedback_score,
        attachedFunctions: workflows.attached_functions,
        modelIdentifier: workflows.model_identifier,
        authContext: workflows.auth_context,
        context: workflows.context,
      })
      .from(workflows)
      .where(and(eq(workflows.cluster_id, clusterId), eq(workflows.id, runId))),
    lastAgentMessage({ clusterId, runId }),
    getWorkflowMetadata({ clusterId, runId }),
  ]);

  return {
    ...workflow,
    metadata,
    // Current a workflow can have multiple "results".
    // For now, we just use the last result.
    // In the future, we will actually persist the workflow result.
    result: agentMessage?.data?.result ?? null,
  };
};

export const addMessageAndResume = async ({
  user,
  id,
  clusterId,
  runId,
  message,
  type,
  metadata,
  skipAssert,
}: {
  user?: Auth;
  id: string;
  clusterId: string;
  runId: string;
  message: string;
  type: "human" | "template" | "supervisor";
  metadata?: RunMessageMetadata;
  skipAssert?: boolean;
}) => {
  if (!skipAssert) await assertRunReady({ clusterId, runId });
  await upsertRunMessage({
    user,
    clusterId,
    runId,
    data: {
      message,
    },
    type,
    id,
    metadata,
  });

  await resumeRun({
    clusterId,
    id: runId,
  });
};

export const resumeRun = async (input: Pick<Run, "id" | "clusterId">) => {
  if (env.NODE_ENV === "test") {
    logger.warn("Skipping run resume. NODE_ENV is set to 'test'.");
    return;
  }

  const sqsResult = await sqs.sendMessage({
    QueueUrl: env.SQS_RUN_PROCESS_QUEUE_URL,
    MessageBody: JSON.stringify({
      runId: input.id,
      clusterId: input.clusterId,
      ...injectTraceContext(),
    }),
  });

  logger.info("Added run processing job to queue", {
    messageId: sqsResult.MessageId,
  });
};

export const generateRunName = async (run: Run, content: string) => {
  if (env.NODE_ENV === "test") {
    logger.warn("Skipping run resume. NODE_ENV is set to 'test'.");
    return;
  }

  if (run.name) {
    logger.info("Skipping run name generation. Name already set.", {
      runId: run.id,
      name: run.name,
    });
    return;
  }

  const sqsResult = await sqs.sendMessage({
    QueueUrl: env.SQS_RUN_GENERATE_NAME_QUEUE_URL,
    MessageBody: JSON.stringify({
      runId: run.id,
      clusterId: run.clusterId,
      content,
      ...injectTraceContext(),
    }),
  });

  logger.info("Added name generation job to queue", {
    runId: run.id,
    messageId: sqsResult.MessageId,
  });
};

export const createRunWithMessage = async ({
  user,
  clusterId,
  message,
  systemPrompt,
  type,
  name,
  test,
  testMocks,
  messageMetadata,
  resultSchema,
  metadata,
  attachedFunctions,
  configId,
  configVersion,
  reasoningTraces,
  interactive,
  enableSummarization,
  modelIdentifier,
  onStatusChange,
  customerAuthToken,
  authContext,
  context,
}: {
  user?: Auth;
  clusterId: string;
  message: string;
  systemPrompt?: string;
  type: "human" | "template";
  name?: string;
  test?: boolean;
  testMocks?: Record<
    string,
    {
      output: Record<string, unknown>;
    }
  >;
  messageMetadata?: RunMessageMetadata;
  onStatusChange?: string;
  resultSchema?: unknown;
  metadata?: Record<string, string>;
  attachedFunctions?: string[];
  configId?: string;
  configVersion?: number;
  reasoningTraces?: boolean;
  interactive?: boolean;
  enableSummarization?: boolean;
  modelIdentifier?: ChatIdentifiers;
  customerAuthToken?: string;
  authContext?: unknown;
  context?: unknown;
}) => {
  const workflow = await createRun({
    user,
    clusterId,
    name,
    test,
    testMocks,
    systemPrompt,
    onStatusChange,
    attachedFunctions,
    resultSchema,
    metadata,
    configId,
    configVersion,
    reasoningTraces,
    interactive,
    enableSummarization,
    modelIdentifier,
    customerAuthToken,
    authContext,
    context,
  });

  await addMessageAndResume({
    id: ulid(),
    user,
    clusterId,
    runId: workflow.id,
    message,
    type,
    metadata: messageMetadata,
    skipAssert: true,
  });

  await generateRunName(workflow, message);

  return workflow;
};

export const assertRunReady = async (input: {
  runId: string;
  clusterId: string;
}) => {
  const run = await getWorkflow(input);
  if (!run) {
    throw new NotFoundError("Run not found");
  }

  logger.info("Asserting run is ready", {
    runId: run.id,
    status: run.status,
  });

  if (!run.interactive) {
    throw new BadRequestError(
      "Run is not interactive and cannot accept new messages.",
    );
  }

  const acceptedStatuses = ["done", "failed", "pending", "paused"];
  if (!acceptedStatuses.includes(run.status ?? "")) {
    throw new RunBusyError(`Run is not ready for new messages: ${run.status}`);
  }

  const [lastMessage] = await getWorkflowMessages({
    clusterId: run.clusterId,
    runId: run.id,
    last: 1,
  });

  if (!lastMessage) {
    return;
  }

  if (lastMessage.type === "agent") {
    // Only Agent messages without function calls are considered ready
    if (!hasInvocations(lastMessage)) {
      return;
    }
  }

  logger.info("Run has unprocessed messages. Workflow will be resumed.", {
    status: run.status,
  });

  await resumeRun({
    clusterId: run.clusterId,
    id: run.id,
  });

  throw new RunBusyError(
    "Run is not ready for new messages: Unprocessed messages",
  );
};

export const getWaitingJobIds = async ({
  clusterId,
  runId,
}: {
  clusterId: string;
  runId: string;
}) => {
  const waitingJobs = await db
    .select({
      id: jobs.id,
    })
    .from(jobs)
    .where(
      and(
        eq(jobs.workflow_id, runId),
        eq(jobs.cluster_id, clusterId),
        or(
          inArray(jobs.status, ["pending", "running"]),
          and(eq(jobs.approval_requested, true), ne(jobs.approved, true)),
        ),
      ),
    );

  return waitingJobs.map((job) => job.id);
};

export const getRunConfigMetrics = async ({
  clusterId,
  configId,
}: {
  clusterId: string;
  configId: string;
}) => {
  return db
    .select({
      createdAt: workflows.created_at,
      count: countDistinct(workflows.id).as("count"),
      feedbackScore: workflows.feedback_score,
      jobCount: countDistinct(jobs.id).as("job_count"),
      jobFailureCount:
        sql<number>`COUNT(${jobs.id}) FILTER (WHERE ${jobs.status} = 'failure')`.as(
          "job_failure_count",
        ),
      timeToCompletion: sql<number>`
        EXTRACT(EPOCH FROM (
          MAX(${workflowMessages.created_at}) - MIN(${workflowMessages.created_at})
        ))
      `.as("time_to_completion"),
    })
    .from(workflows)
    .leftJoin(jobs, eq(workflows.id, jobs.workflow_id))
    .leftJoin(workflowMessages, eq(workflows.id, workflowMessages.workflow_id))
    .where(
      and(
        eq(workflows.cluster_id, clusterId),
        eq(workflows.config_id, configId),
      ),
    )
    .groupBy(workflows.id, workflows.created_at, workflows.feedback_score)
    .limit(1000);
};

export const createRetry = async ({
  clusterId,
  runId,
  messageId,
}: {
  clusterId: string;
  runId: string;
  messageId: string;
}) => {
  const { deleted } = await prepMessagesForRetry({
    clusterId,
    runId,
    messageId,
  });

  await db
    .update(workflows)
    .set({
      status: "pending",
      failure_reason: null,
    })
    .where(eq(workflows.id, runId));

  await resumeRun({
    clusterId,
    id: runId,
  });

  return {
    deleted,
  };
};

export const getRunCustomerAuthToken = async ({
  clusterId,
  runId,
}: {
  clusterId: string;
  runId: string;
}) => {
  const [workflow] = await db
    .select({
      customerAuthToken: workflows.customer_auth_token,
    })
    .from(workflows)
    .where(and(eq(workflows.id, runId), eq(workflows.cluster_id, clusterId)))
    .limit(1);

  if (!workflow) {
    throw new NotFoundError("Run not found");
  }

  return workflow.customerAuthToken;
};
