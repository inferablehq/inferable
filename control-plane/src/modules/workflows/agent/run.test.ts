import { Run } from "../workflows";
import { buildMockTools, findRelevantTools } from "./run";
import { upsertServiceDefinition } from "../../service-definitions";
import { createOwner } from "../../test/util";
import { ulid } from "ulid";

describe("findRelevantTools", () => {
  it("should return explicitly attached tools", async () => {
    const owner = await createOwner();
    await upsertServiceDefinition({
      service: "testService",
      definition: {
        name: "testService",
        functions: [
          {
            name: "someFunction",
            schema: mockTargetSchema,
          },
          {
            name: "someOtherFunction",
            schema: mockTargetSchema,
          },
        ],
      },
      owner,
    });

    const workflow = {
      id: Math.random().toString(36).substring(2),
      clusterId: owner.clusterId,
      status: "running" as const,
      attachedFunctions: ["testService_someFunction"],
    };

    const tools = await findRelevantTools({
      workflow,
      messages: [
        {
          data: {
            message: "Call someFunction",
          },
          type: "human" as const,
          clusterId: owner.clusterId,
          runId: "test-workflow-id",
          id: ulid(),
        },
      ],
      status: "running",
      waitingJobs: [],
      allAvailableTools: [],
    });

    expect(tools.map((tool) => tool.name)).toContain(
      "testService_someFunction",
    );
    expect(tools.map((tool) => tool.name)).not.toContain(
      "testService_someOtherFunction",
    );
  });
});

const mockTargetSchema = JSON.stringify({
  type: "object",
  properties: {
    test: {
      type: "string",
    },
  },
});

describe("buildMockTools", () => {
  let workflow: Run;

  const service = "testService";
  beforeAll(async () => {
    const owner = await createOwner();
    await upsertServiceDefinition({
      service,
      definition: {
        name: service,
        functions: [
          {
            name: "someFunction",
            schema: mockTargetSchema,
          },
        ],
      },
      owner,
    });

    workflow = {
      id: Math.random().toString(36).substring(2),
      clusterId: owner.clusterId,
      status: "running",
    };
  });

  it("should return an empty object if no mocks are defined", async () => {
    const tools = await buildMockTools(workflow);
    expect(tools).toEqual({});
  });

  it("should return an empty object if test is not enabled", async () => {
    const tools = await buildMockTools({
      ...workflow,
      testMocks: {
        testService_someFunction: {
          output: {
            foo: "bar",
          },
        },
      },
    });
    expect(tools).toEqual({});
  });

  it("should return mock tools", async () => {
    const tools = await buildMockTools({
      ...workflow,
      test: true,
      testMocks: {
        testService_someFunction: {
          output: {
            foo: "bar",
          },
        },
      },
    });
    expect(Object.keys(tools)).toEqual(["testService_someFunction"]);

    const result = await tools["testService_someFunction"].func({ test: "" });
    expect(JSON.parse(result)).toEqual({
      result: {
        foo: "bar",
      },
      resultType: "resolution",
      status: "success",
    });
  });
});
