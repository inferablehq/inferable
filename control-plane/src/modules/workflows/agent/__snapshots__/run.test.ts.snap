// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatJobsContext should handle unparseable results 1`] = `
"<previous_jobs status="failed">
    <input>"<string>"</input>
    <output>"<string>"</output>
<input>"<string>"</input>
    <output>"<string>"</output>
<input>"<number>"</input>
    <output>"<number>"</output>
  </previous_jobs>"
`;
